### DISCUSSION TITLE
(PDF-)Printing in WxMaxima on DIN-A4 looses lines

### DISCUSSION THREAD

**Discussion by hbiber** [wxMaxima-developers/wxmaxima#1871] [General]
We're using WxMaxima within the "Prüfungsstick" (see: https://www.bfh.ch/de/forschung/forschungsbereiche/lernstick/ ) for teaching and several exams in our math courses.
At the end of an exam the wxmx-file of the pupil must be printed from a Windows-PC (Win10 or Win11) on  paper with DIN-A4-pagesize. 
However printed directly to a laserprinter (Kyocera, HP, ...) with A4-paper or the "Windows PDF-Printer" (configured for A4-Paper) some lines of the wxmx-file are missing. Blocks with "normal" Text are truncated, input and output-lines are missing.
I've played with the "printout settings" (-> print margins) but no success.
(my system: Win11, Wxmaxima 23.08.0, Maxima version="5.47.0")

To reproduce this effect just create an wxmx-file with several input-/ouput-lines and text-fields.

It would be a great help (or feature ;-) ) if no lines are missing or if I could set the pagesize (here DIN-A4) in the "printout settings".

Thanks for any hints

**Comment by gunterkoenigsmann** [comment]
Making them use printing in a program that never popperly supported printing and doing this in an exam situation sounds like we germans treat students...

Will give printing another try (actually have started that with commits afd170b769783dd9af60f8df2289ac9acf50a1b0, e14cd3cef2818dfccf9e58af9acef04f78fda072, 9562649cefc424b7c149252055cc42625f9521b8 and 1e2357c2e4c404f948cb825f82f5ed84730c2519). But I cannot promise anything: until now I never managed to get printing in wxMaxima into a working order. 

Once I have something you can test I'll tell you: Printing in wxWigets effectively results in a bunch of png images in an PDF container. I have no idea if we can write any automatic tests for such a thing...

**Comment by gunterkoenigsmann** [comment]
Ok... ...I have now

- restructured the display code so it shares as many code paths with printing, as possible: This way if printing breaks again we are more likely to see it in time.
- simplified the display and printout routines to a point that I am actually able to understand them.
- corrected a few off-by-one errors
- removed two loop conditions that didn't help
- and found out why printing only worked every other time,

-tomorrow the windows installer at https://wolfgang.dautermann.at/maxima/nightlybuild/ will contain a wxMaxima that contains all those fixes. Ifit works for you I can make an official wxMaxima release. How to get that one into the Prüfungsstick I don't know, though.

**Comment by hbiber** [comment]
Hallo Gunter,
first of all, happy new year and best wishes from Löhne (Northrhein-Westfalia).
Thanks in advance for your work. I will give it a try and report any things.
It's not necessary to include it in the "Prüfungsstick". It will be just for the "teachers PC/notebook".
For your info how we'll use wxMaxima:
* In an exam (Mathematik-Leistungskurs) the pupils work between two and three an a half hours with wxmaxima (under Linux/Prüfungsstick) to solve the tasks.
* When they finish they power down their notebooks and take the USB-Stick to the teacher who now opens the wxmaxima-file (*.wxmx) under Win10/Win11 and print it directly to the local printer (laserprinter from Kyocera, HP or Lexmark).
bye and until tomorrow ;-)
Holger
Ok... ...I have now
restructured the display code so it shares as many code paths with printing, as possible: This way if printing breaks again we are more likely to see it in time.
simplified the display and printout routines to a point that I am actually able to understand them.
corrected a few off-by-one errors
removed two loop conditions that didn't help
and found out why printing only worked every other time,
-tomorrow the windows installer at https://wolfgang.dautermann.at/maxima/nightlybuild/ <https://wolfgang.dautermann.at/maxima/nightlybuild/> will contain a wxMaxima that contains all those fixes. Ifit works for you I can make an official wxMaxima release. How to get that one into the Prüfungsstick I don't know, though.

**Reply by gunterkoenigsmann** [reply]
Richard has just informed me that there is a chance that this night the nightly build will fail. Ergo I cannot guarantee that tomorrow's version will already work.
